File: backend/models/Beacon.js
const mongoose = require('mongoose');

const BeaconSchema = new mongoose.Schema(
  {
    beaconId: {
      type: String,
      required: [true, 'Beacon identifier is required (e.g. BEACON-GF-01)'],
      unique: true,
      trim: true,
      index: true
    },
    buildingId: {
      type: String,
      required: [true, 'Associated buildingId is required'],
      index: true
    },
    floorId: {
      type: Number,
      required: [true, 'Floor number is required (0 for Ground, 1 for Floor 1, etc.)'],
      default: 0
    },
    nodeId: {
      type: String,
      required: [true, 'Associated navigation NavNode id is required (e.g. univ_gate)'],
      index: true
    },
    x: {
      type: Number,
      required: [true, 'X coordinate on normalized 0..1000 canvas is required'],
      default: 500
    },
    y: {
      type: Number,
      required: [true, 'Y coordinate on normalized 0..1000 canvas is required'],
      default: 500
    },
    major: {
      type: Number,
      default: 1
    },
    minor: {
      type: Number,
      default: 1
    },
    txPower: {
      type: Number,
      default: -59 // Measured RSSI at 1 meter in dBm
    },
    activeStatus: {
      type: Boolean,
      default: true
    },
    metadata: {
      macAddress: { type: String, default: '' },
      batteryLevel: { type: Number, default: 100 },
      description: { type: String, default: 'PathFinder BLE Anchor Beacon' }
    }
  },
  {
    timestamps: true
  }
);

module.exports = mongoose.models.Beacon || mongoose.model('Beacon', BeaconSchema);

File: backend/app/src/main/java/com/example/data/model/Beacon.kt
package com.example.data.model

data class Beacon(
    val beaconId: String,
    val buildingId: String,
    val floorId: Int = 0,
    val nodeId: String = "",
    val x: Float = 0f, // Coordinates on normalized 0..1000 map canvas
    val y: Float = 0f,
    val major: Int = 1,
    val minor: Int = 1,
    val txPower: Int = -59, // Calibrated RSSI at 1 meter in dBm
    val activeStatus: Boolean = true,
    val macAddress: String = "",
    val description: String = "",
    val name: String = "",
    val pathLossExponent: Float = 2.2f
) {
    val floor: Int get() = floorId
}

enum class PositionConfidence {
    HIGH,
    MEDIUM,
    LOW;

    val label: String
        get() = name.lowercase().replaceFirstChar { if (it.isLowerCase()) it.titlecase() else it.toString() }
}

enum class PositionSource(val label: String) {
    GPS("GPS Outdoor"),
    QR_CALIBRATION("QR Ground Truth"),
    BLE_BEACON("BLE Trilateration"),
    PDR_SENSOR("PDR Step & Compass"),
    MANUAL("Manual Anchor"),
    SIMULATION("Simulation Mode")
}

data class IndoorPosition(
    val buildingId: String,
    val floorId: Int,
    val x: Float,
    val y: Float,
    val nearestNodeId: String? = null,
    val headingDegrees: Float = 0f,
    val confidence: PositionConfidence = PositionConfidence.HIGH,
    val confidenceScore: Float = 1.0f, // 0.0 .. 1.0
    val source: PositionSource = PositionSource.QR_CALIBRATION,
    val activeBeaconsCount: Int = 0,
    val timestamp: Long = System.currentTimeMillis(),
    val accuracyMeters: Float = 1.5f,
    val isSnapped: Boolean = false,
    val snappedNodeId: String? = null
) {
    val floor: Int get() = floorId
    val provider: PositionSource get() = source
}
